exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/decoder/Packet.java

```diff
@@ -0,0 +1,371 @@
+import static org.apache.drill.exec.store.pcap.Utils.getShort;
+
+public class Packet {
+  // pcap header
+  //        typedef struct pcaprec_hdr_s {
+  //            guint32 ts_sec;         // timestamp seconds
+  //            guint32 ts_usec;        // timestamp microseconds */
+  //            guint32 incl_len;       // number of octets of packet saved in file */
+  //            guint32 orig_len;       // actual length of packet */
+  //        } pcaprec_hdr_t;
+  private long timestamp;
+  private int originalLength;
+
+  private byte[] raw;
+
+  private int etherOffset;
+  private int ipOffset;
+
+  private int packetLength;
+  private int etherProtocol;
+  private int protocol;
+
+  private boolean isRoutingV6;
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean readPcap(final InputStream in, final boolean byteOrder, final int maxLength) throws IOException {
+    byte[] pcapHeader = new byte[PacketConstants.PCAP_HEADER_SIZE];
+    int n = in.read(pcapHeader);
+    if (n < pcapHeader.length) {
+      return false;
+    }
+    decodePcapHeader(pcapHeader, byteOrder, maxLength, 0);
+
+    raw = new byte[originalLength];
+    n = in.read(raw);
+    if (n < 0) {
+      return false;
+    }
+    etherOffset = 0;
+
+    decodeEtherPacket();
+    return true;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public int decodePcap(final byte[] buffer, final int offset, final boolean byteOrder, final int maxLength) {
+    raw = buffer;
+    etherOffset = offset + PacketConstants.PCAP_HEADER_SIZE;
+    decodePcapHeader(raw, byteOrder, maxLength, offset);
+    decodeEtherPacket();
+    return offset + PacketConstants.PCAP_HEADER_SIZE + originalLength;
+  }
+
+  public String getPacketType() {
+    if (isTcpPacket()) {
+      return "TCP";
+    } else if (isUdpPacket()) {
+      return "UDP";
+    } else if (isArpPacket()) {
+      return "ARP";
+    } else if (isIcmpPacket()) {
+      return "ICMP";
+    } else {
+      return "unknown";
+    }
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isIpV4Packet() {
+    return etherProtocol == PacketConstants.IPv4_TYPE;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isIpV6Packet() {
+    return etherProtocol == PacketConstants.IPv6_TYPE;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isPPPoV6Packet() {
+    return etherProtocol == PacketConstants.PPPoV6_TYPE;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isTcpPacket() {
+    return protocol == PacketConstants.TCP_PROTOCOL;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isUdpPacket() {
+    return protocol == PacketConstants.UDP_PROTOCOL;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isArpPacket() {
+    return protocol == PacketConstants.ARP_PROTOCOL;
+  }
+
+  @SuppressWarnings("WeakerAccess")
+  public boolean isIcmpPacket() {
+    return protocol == PacketConstants.ICMP_PROTOCOL;
+  }
+
+  public long getSessionHash() {
```

Comment:
This is very useful. It is the way that TCP sessions are reassembled (which is a big deal).